### DISCUSSION TITLE
Extending destack to integrate with opensource headless CMS

### DISCUSSION THREAD

**Discussion by wenhao-mui** [LiveDuo/destack#113] [Ideas]
There are lots of opensource headless CMS (e.g.: umbraco, strapi) where their frontend rendering are tightly couple. I was planning to build a website builder and WYSWYG features and integrate with these CMS. Happened to come across destack and it already fulfil most of the website builder and WYSWYG.

I'm planning create a modular plugin on top of destack to integrate with various opensource, headless CMS. Rather than storing the output into the version control, I will store them in CMS. In short, I'm trying to replicate some of the enterprise CMS features but make opensource available XD

**Comment by LiveDuo** [comment]
@wenhao-mui Hey Wen,

Glad to hear you found it useful. I have been looking into integrating a CMS for quite a while now and never got to build it.

There are multiple ways to do it. I'd probably go with client-side templates (similar to how HTMX handle this same problem https://htmx.org/extensions/client-side-templates/).

If you build something interesting, feel free to share it here. I'd be very curious to check it out.

**Comment by sreekanthedayar** [comment]
Does this had any updates. Because I just saw while searching for React based page builders. We are having a use case for templating but the data should be fetched from content API. eg: suppose i have a content called invoice with defined field and values. I then create attach to a template and map fields with the invoice content. Then the template should fetch the data from invoice content. .

**Reply by LiveDuo** [reply]
@sreekanthedayar There isn't a built in way to do this right now but it shouldn't be too hard. The generated HTML is injected [here](https://github.com/LiveDuo/destack/blob/main/lib/client/vanilla/index.tsx#L32). You can have some templating language, fetch the API data with a useEffect and then just replace the fields with the values in the HTML string ie. the `_templateData.content` variable.

**Reply by sreekanthedayar** [reply]
That's great. I will keep posted. We are planning to start this feature after couple of months.
